File: back-end/app/models/portfolio.py
"""Pydantic models for portfolio."""

from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class PortfolioItem(BaseModel):
    """A position in the portfolio."""
    id: Optional[str] = None
    symbol: str
    bucket: Optional[str] = None          # SAFE_INCOME, HIGH_RISK
    account_type: Optional[str] = None    # TFSA, RRSP, TAXABLE
    shares: Optional[Decimal] = None
    avg_cost: Optional[Decimal] = None
    currency: str = "CAD"                 # CAD, USD
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class PortfolioAddRequest(BaseModel):
    """Request to add a position."""
    symbol: str = Field(..., min_length=1, max_length=10, pattern=r"^[A-Z0-9.\-]+$")
    bucket: Optional[str] = Field(None, pattern=r"^(SAFE_INCOME|HIGH_RISK)$")
    account_type: Optional[str] = Field(None, pattern=r"^(TFSA|RRSP|TAXABLE)$")
    shares: Optional[Decimal] = Field(None, gt=0)
    avg_cost: Optional[Decimal] = Field(None, gt=0)
    currency: str = Field("CAD", pattern=r"^(CAD|USD)$")


class PortfolioUpdateRequest(BaseModel):
    """Request to update a position."""
    bucket: Optional[str] = Field(None, pattern=r"^(SAFE_INCOME|HIGH_RISK)$")
    account_type: Optional[str] = Field(None, pattern=r"^(TFSA|RRSP|TAXABLE)$")
    shares: Optional[Decimal] = Field(None, gt=0)
    avg_cost: Optional[Decimal] = Field(None, gt=0)
    currency: Optional[str] = Field(None, pattern=r"^(CAD|USD)$")


class PortfolioResponse(BaseModel):
    """Portfolio list response."""
    items: list[PortfolioItem]
    count: int

File: front-end/src/types/portfolio.ts
export interface PortfolioItem {
  id: string
  symbol: string
  bucket: 'SAFE_INCOME' | 'HIGH_RISK' | null
  account_type: 'TFSA' | 'RRSP' | 'TAXABLE' | null
  shares: number | null
  avg_cost: number | null
  currency: 'CAD' | 'USD'
  created_at: string
  updated_at: string
}

export interface PortfolioResponse {
  items: PortfolioItem[]
  count: number
}

export interface PortfolioAddRequest {
  symbol: string
  bucket?: 'SAFE_INCOME' | 'HIGH_RISK' | null
  account_type?: 'TFSA' | 'RRSP' | 'TAXABLE' | null
  shares?: number
  avg_cost?: number
  currency?: 'CAD' | 'USD'
}

export interface PortfolioUpdateRequest {
  bucket?: 'SAFE_INCOME' | 'HIGH_RISK' | null
  account_type?: 'TFSA' | 'RRSP' | 'TAXABLE' | null
  shares?: number
  avg_cost?: number
  currency?: 'CAD' | 'USD'
}

export interface Position {
  id: string
  symbol: string
  entry_price: number
  entry_date: string
  shares: number
  account_type: 'TFSA' | 'RRSP' | 'TAXABLE' | null
  bucket: 'SAFE_INCOME' | 'HIGH_RISK' | null
  currency: 'CAD' | 'USD'
  target_price: number | null
  stop_loss: number | null
  notes: string | null
  status: 'OPEN' | 'CLOSED' | 'STOPPED_OUT'
  exit_price: number | null
  exit_date: string | null
  exit_reason: 'USER_CLOSE' | 'TARGET_HIT' | 'STOP_HIT' | 'SIGNAL_WEAKENED' | null
  pnl_amount: number | null
  pnl_percent: number | null
  last_signal_score: number | null
  last_signal_status: string | null
  created_at: string
  updated_at: string
}

export interface PositionsResponse {
  positions: Position[]
  count: number
}

export interface PositionOpenRequest {
  symbol: string
  entry_price: number
  shares: number
  account_type?: 'TFSA' | 'RRSP' | 'TAXABLE' | null
  bucket?: 'SAFE_INCOME' | 'HIGH_RISK' | null
  currency?: 'CAD' | 'USD'
  target_price?: number
  stop_loss?: number
  notes?: string
}

export interface PositionUpdateRequest {
  target_price?: number
  stop_loss?: number
  notes?: string
}

export interface PositionCloseRequest {
  exit_price: number
}
